### PR TITLE
Detail tabs

### PR DESCRIPTION
**Pull request by sunglam**
#minor#

# CHANGELOG
- Added detail tabs ( user can choose and add more tabs from a list of tabs).
- A detail tab can have vertical sub tabs with sub tab's content area and related actions in vertical action bar.

### CODE REVIEW

File: ember-cli-build.js [hunk 1]
```diff
@@ -1,8 +1,11 @@
 /* eslint-env node */
-const EmberAddon = require('ember-cli/lib/broccoli/ember-addon')
+const EmberApp = require('ember-cli/lib/broccoli/ember-addon')
```
Comment (juwara0) [inline ember-cli-build.js]: Please revert this change. This is an `EmberAddon` not an `EmberApp`.

File: ember-cli-build.js [hunk 2]
```diff
@@ -1,8 +1,11 @@
...
 
 module.exports = function (defaults) {
-  var app = new EmberAddon(defaults, {
+  var app = new EmberApp(defaults, {
```
Comment (juwara0) [inline ember-cli-build.js]: `EmberAddon` instead of `EmberApp`.

File: package.json [hunk 3]
```diff
@@ -48,10 +48,14 @@
...
-    "ember-export-application-global": "^1.0.5",
+    "ember-disable-proxy-controllers": "^1.0.1",
+    "ember-element-resize-detector": "0.1.5",
+    "ember-export-application-global": "^1.1.1",
+    "ember-frost-info-bar": "8.3.5",
     "ember-hook": "^1.4.1",
-    "ember-load-initializers": "^0.6.0",
+    "ember-load-initializers": "0.6.3",
+    "ember-math-helpers": "2.1.0",
+    "ember-elsewhere": "1.0.1",
```
Comment (juwara0) [inline package.json]: This got moved but is now not in alphabetical order. Please alphabetize.

File: package.json [hunk 4]
```diff
@@ -48,10 +48,14 @@
     "ember-computed-decorators": "0.3.0",
     "ember-concurrency": "0.7.19",
     "ember-disable-prototype-extensions": "^1.1.0",
-    "ember-elsewhere": "1.0.1",
-    "ember-export-application-global": "^1.0.5",
+    "ember-disable-proxy-controllers": "^1.0.1",
+    "ember-element-resize-detector": "0.1.5",
+    "ember-export-application-global": "^1.1.1",
```
Comment (juwara0) [inline package.json]: Please leave this at: `"ember-export-application-global": "^1.0.5",`

File: package.json [hunk 5]
```diff
@@ -48,10 +48,14 @@
     "ember-computed-decorators": "0.3.0",
     "ember-concurrency": "0.7.19",
     "ember-disable-prototype-extensions": "^1.1.0",
-    "ember-elsewhere": "1.0.1",
-    "ember-export-application-global": "^1.0.5",
+    "ember-disable-proxy-controllers": "^1.0.1",
```
Comment (juwara0) [inline package.json]: This should not be necessary since this has been gone since Ember 2.0: https://github.com/cibernox/ember-disable-proxy-controllers#ember-disable-proxy-controllers

File: tests/dummy/app/pods/demo/more-detail-tabs/controller.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+  // State management
+
+  tabs: A([
+    'View',
+    'Attributes',
+    'Footab',
+    'Bartab',
+    'Biztab'
+    // 'Baztab',
```
Comment (juwara0) [inline tests/dummy/app/pods/demo/more-detail-tabs/controller.js]: Did you mean to leave these commented out or did you intend to remove these lines?

File: tests/dummy/app/pods/demo/more-detail-tabs/template.hbs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+          tabActions=tabActions
+          onDispatch=(action 'dispatch')
+        }}
+      </div>
+    </div>
+    {{! END-SNIPPET }}
+  </div>
+</div>
+
+
```
Comment (juwara0) [inline tests/dummy/app/pods/demo/more-detail-tabs/template.hbs]: Please remove the extra blank lines.

File: tests/integration/components/frost-detail-subtab-test.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
+import {expect} from 'chai'
+import {$hook, initialize} from 'ember-hook'
```
Comment (juwara0) [inline tests/integration/components/frost-detail-subtab-test.js]: Instead of `initialize`, name it `initializeHook`so that it is known what is being initialized. Please make this change in all occurrences.

File: tests/integration/components/frost-detail-subtabs-test.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        expect(this.$('.frost-detail-subtab.selected')).to.have.length(1)
+      })
+  })
+
+  it('should have two subtabs', function () {
+    this.render(template)
+
+    return wait()
+      .then(() => {
+        expect(this.$('.frost-detail-subtab')).to.have.length(2)
```
Comment (juwara0) [inline tests/integration/components/frost-detail-subtabs-test.js]: Would it make sense for there to be a `ember-hook` at this level as well? If so, apply the hook to other places as appropriate in this test suite.

File: tests/integration/components/frost-detail-tabs-test.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    this.render(template)
+
+    return wait()
+      .then(() => {
+        expect($hook(`${myHook}-${tabs[0].id}`)).to.have.length(1)
+        expect($hook(`${myHook}-${tabs[1].id}`)).to.have.length(1)
+      })
+  })
+
+  it('should have slected tab  by default', function () {
```
Comment (juwara0) [inline tests/integration/components/frost-detail-tabs-test.js]: Spelling: `slected` should be `selected`.

File: addon/components/frost-detail-subtabs.js [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+      pack: 'frost'
+    }
+  },
+
+  @readOnly
+  @computed('subtabs.[]')
+  _subtabs (subtabs) {
+    if (isEmpty(subtabs) || typeOf(subtabs[0]) === 'object' || typeOf(subtabs[0]) === 'instance') {
+      // const isMissingProperties = subtabs.some(({id, label, icon, pack}) => {
+      //   return !id || !label
```
Comment (juwara0) [inline addon/components/frost-detail-subtabs.js]: Please remove the commented code if it is not needed.

File: addon/components/frost-detail-subtabs.js [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    return {
+      // options
+
+      // state
+    }
+  },
+
+  // == Computed Properties ===================================================
+
+  // Wrap selectedTab for the same reasons outlined in the comments for _tabs
```
Comment (juwara0) [inline addon/components/frost-detail-subtabs.js]: Is the comment for `_tabs` that this comment references still in the code base? What happens if that comment were to be deleted or is already deleted? Instead, it would be better to add the comment block here as well.

File: addon/components/frost-detail-tab.js [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+      return false
+    }
+
+    const tabLeft = this.$().position().left
+    const tabWidth = this.$().outerWidth()
+    const value = 0
+    // This is annoying, but the math appears to be slightly off, maybe because of the border
+    // between the tabs - so we're dropping one off the tabLeft for the border.
+    // Why the viewport needs another pixel for this to calculate correctly, I don't know...
+    // TODO Find out why...
```
Comment (juwara0) [inline addon/components/frost-detail-tab.js]: Please put your username and date against this TODO and open an issue as well.

File: addon/components/frost-detail-tabs-more-outlet.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+import layout from '../templates/components/frost-detail-tabs-more-outlet'
+
+export default Component.extend({
+
+  // == Dependencies ==========================================================
+
+  // == Keyword Properties ====================================================
+
+  layout,
+  tagName: '',
```
Comment (juwara0) [inline addon/components/frost-detail-tabs-more-outlet.js]: Can this be changed to a `span` maybe?

**Review by juwara0 — CHANGES_REQUESTED** [review]

File: ember-cli-build.js (hunk 1 above)
Comment (sunglam) [reply ember-cli-build.js]: Sorry ..big mistake ..  Thanks for this comment.. I is done

File: tests/integration/components/frost-detail-subtab-test.js (hunk 8 above)
Comment (sunglam) [reply tests/integration/components/frost-detail-subtab-test.js]: Good idea . Thanks .. It is done.

File: tests/integration/components/frost-detail-subtabs-test.js (hunk 9 above)
Comment (sunglam) [reply tests/integration/components/frost-detail-subtabs-test.js]: Good idea . Tnx . It is done.

File: tests/integration/components/frost-detail-tabs-test.js (hunk 10 above)
Comment (sunglam) [reply tests/integration/components/frost-detail-tabs-test.js]: done.

File: addon/components/frost-detail-subtabs.js (hunk 11 above)
Comment (sunglam) [reply addon/components/frost-detail-subtabs.js]: Done

File: addon/components/frost-detail-tab.js (hunk 13 above)
Comment (sunglam) [reply addon/components/frost-detail-tab.js]: Done

File: addon/components/frost-detail-tabs-more-outlet.js (hunk 14 above)
Comment (sunglam) [reply addon/components/frost-detail-tabs-more-outlet.js]: Done

File: ember-cli-build.js (hunk 2 above)
Comment (sunglam) [reply ember-cli-build.js]: Done

File: package.json (hunk 5 above)
Comment (sunglam) [reply package.json]: Done

File: package.json (hunk 4 above)
Comment (sunglam) [reply package.json]: Done

File: package.json (hunk 3 above)
Comment (sunglam) [reply package.json]: Done

File: tests/dummy/app/pods/demo/more-detail-tabs/controller.js (hunk 6 above)
Comment (sunglam) [reply tests/dummy/app/pods/demo/more-detail-tabs/controller.js]: Done

File: tests/dummy/app/pods/demo/more-detail-tabs/template.hbs (hunk 7 above)
Comment (sunglam) [reply tests/dummy/app/pods/demo/more-detail-tabs/template.hbs]: Done

File: addon/components/frost-detail-subtabs.js (hunk 12 above)
Comment (sunglam) [reply addon/components/frost-detail-subtabs.js]: Done

**Comment by rox163** [comment]
:+1:

[Approved with PullApprove](https://pullapprove.com/ciena-frost/ember-frost-tabs/pull-request/66/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ciena-frost/ember-frost-tabs)

**Review by juwara0 — APPROVED** [review]
